src/nupic/types/ptr_types.hpp

```diff
@@ -0,0 +1,16 @@
+#ifndef NTA_PTR_TYPES_HPP
+#define NTA_PTR_TYPES_HPP
+
+#include <memory>
+
+namespace nupic
+{
+    class Link;
+    class Region;
+
+    typedef std::shared_ptr<Link> Link_Ptr_t;
```

Comment:
(not important, just sayin') is this worth? `shared_ptr<Link>` feels more self-explaining

Reply:
Actually, you may be right.  This ptr_types.hpp file was from @chhenning 's work in setting up pybind11.  I did not intent this file to be in this PR but I jumped the gun a little.  I think I will go back to putting in std::shared_ptr<Link> and std::shared_ptr<Region>.  SWIG is having problems with Link_Ptr_t for some reason (says its undefined) and it is not worth the time to figure out why.

Reply:
I had the same problem in the other PR, https://github.com/htm-community/nupic.cpp/pull/108/files#diff-393e610699c2531fb47bbb66307d24ecR46 
SWIG does not handle well `<T>` types. 

http://www.acodersjourney.com/2016/05/top-10-dumb-mistakes-avoid-c-11-smart-pointers/ 
This was an interesting read for me, takeaway: When possible, use `unique_ptr` instead of shared.